### PR TITLE
investigate: Support MM tests by using the restart API instead of cloning

### PR DESCRIPTION
**Pull request by Martchus**
See https://progress.opensuse.org/issues/95783

---

This is still a draft to see how it would look like. Restarting jobs with different settings hasn't been implemented yet.

### CODE REVIEW

File: openqa-investigate [hunk 1]
```diff
@@ -24,45 +24,61 @@ jq_output_limit="${jq_output_limit:-15}"
 curl_args=(--user-agent "openqa-investigate")
 echoerr() { echo "$@" >&2; }
 
-clone() {
-    local origin id name_suffix refspec job_data unsupported_cluster_jobs name base_prio clone_settings casedir repo out url clone_id
+declare -a handled_job_ids
+
+restart_job() {
```
Comment (okurz) [inline openqa-investigate:29]: Despite the name of the API I prefer the word "clone" as we are not actually restarting the original job which is immutable but rather really create a clone and it's actually an altered clone. So I suggest to keep the wording "clone". Maybe we find an even better wording for "clone or copy but with changes" or "create new jobs based on existing ones with only small changes". I would say "clone" does not imply that we would use a clone script or something. Can be a /restart API just as well. One more benefit of "clone" vs. "restart" is just 5 characters vs. 7 ;)

**Review by okurz — COMMENTED** [review]
Your change reminds me of https://github.com/os-autoinst/scripts/pull/77 . So do you think it's a good idea for the openQA jobs/$id/restart route to take test parameters for overwriting like we do within the openqa-clone-job script?

File: openqa-investigate [hunk 2]
```diff
@@ -24,39 +24,48 @@ jq_output_limit="${jq_output_limit:-15}"
...
-    local origin id name_suffix refspec job_data unsupported_cluster_jobs name base_prio clone_settings casedir repo out url clone_id
+    local origin id name_suffix refspec job_data unsupported_cluster_jobs name base_prio restart_settings casedir repo out url handled_ids restart_ids
     origin=${1:?"Need 'origin'"}
     id=${2:?"Need 'id'"}
     name_suffix=${3+":$3"}
     refspec=${4+$4}
+
+    # check whether the job has already been handled as part of restarting another job and skip if that's the case
+    local handled_job=${handled_job_ids["$job_id"]}
+    [[ $handled_job ]] && echoerr "skipping job $id: it has already been handled when restarting $handled_job" && return 2
```
[the remark this answers is not in the document]
Comment (okurz) [reply openqa-investigate:38]: Hm, then maybe not output anything at all?

[the remark this answers is not in the document]
Comment (Martchus) [reply openqa-investigate]: Changed it to not output anything at all.

**Review by okurz — COMMENTED** [review]
+1

**Comment by Martchus** [comment]
I gave it some rudimentary local testing; it is still able to create new jobs and the settings are there. Of course now they show up as clones and prevent the original job (with original settings) to be restarted. Not sure whether that's a problem. If it is I'll add another flag to the restart route to avoid that (which should be fairly simple).

**Comment by okurz** [comment]
> I gave it some rudimentary local testing; it is still able to create new jobs and the settings are there. Of course now they show up as clones and prevent the original job (with original settings) to be restarted

So you are saying same as in before with openqa-investigate the original job is not altered, right?

File: openqa-investigate (hunk 1 above)
Comment (kalikiana) [reply openqa-investigate]: I would argue Investigation creates "clones" but also alternate versions of the job. Can we quantify what we think is a clone? What has to be the same?

**Comment by Martchus** [comment]
I'm not quite sure what you're getting. If we'd merge the PR right now the original job's `clone_id` would be set but besides that it'll stay the same. I'm asking whether that's ok or should be avoided.

**Review by okurz — CHANGES_REQUESTED** [review]
We must not cause investigation jobs to be actual clones which would prevent restarting over the webUI. So in openQA we should allow to add an additional parameter to created an altered clone with settings but cut the link to not show up as real clone.

**Comment by Martchus** [comment]
I added `clone=0`, PR for implementing that in openQA: https://github.com/os-autoinst/openQA/pull/4487

**Review by okurz — COMMENTED** [review]
Looks fine but needs an according openQA as change first. I recommend to await deployment on o3 first and conduct it once against that instance manually first

**Comment by Martchus** [comment]
It should be safe to merge now, see https://progress.opensuse.org/issues/95783#note-11

**Review by kalikiana — APPROVED** [review]

**Review by okurz — APPROVED** [review]

**Pull request merged**